system/jlib/jio.cpp

```diff
@@ -1344,6 +1344,106 @@ IRowStream *createConcatRowStream(unsigned numstreams,IRowStream** streams,bool
     }
 }
 
+class CStreamLineReader : public CSimpleInterfaceOf<IStreamLineReader>
+{
+    Linked<ISimpleReadStream> stream;
+    bool preserveEols;
+    size32_t chunkSize = 8192;
+    MemoryBuffer buffer;
+    const char *startPtr = nullptr;
+    const char *endPtr = nullptr;
+    const char *currentPtr = nullptr;
+
+    bool refill()
+    {
+        size32_t r = stream->read(chunkSize, buffer.bufferBase());
```

Comment:
Inside the read(), another buffer is used. So, you use that buffer to fill this buffer - less efficient.

Reply:
The stream (a ISimpleReadStream) may use a buffer or may not, that is implementation dependent.

It will work with any implementation, but given this IStreamLineReader implementation needs to buffer to read ahead, it would be sensible to give a unbuffered ISimpleReadStream implementation to the factory method (createLineReader).

That's why in my example I showed using an unbuffered stream : 
```
    Owned<IFileIOStream> ios = createIOStream(rIO); // unbuffered
    Owned<IStreamLineReader> lineReader = createLineReader(ios, true);
```

ios has no buffer.

Reply:
I see.